crates/compiler/builtins/roc/Dict.roc

```diff
@@ -208,3 +210,413 @@ insertIfVacant = \dict, key, value ->
         dict
     else
         Dict.insert dict key value
+
+# We have decided not to expose the standard roc hashing algorithm.
+# This is to avoid external dependence and the need for versioning.
+# For now, it will just live within the Dict file.
+# If we add non-exposed roc standard library files, this should be moved to it's own file.
+# The current implementation is a form of [Wyhash final3](https://github.com/wangyi-fudan/wyhash/blob/a5995b98ebfa7bd38bfadc0919326d2e7aabb805/wyhash.h).
+# It is 64bit and little endian specific currently.
+LowLevelHasher := { originalSeed : U64, state : U64 } has [
+         Hasher {
+             addBytes,
+             addU8,
+             addU16,
+             addU32,
+             addU64,
+             addU128,
+             addI8,
+             addI16,
+             addI32,
+             addI64,
+             addI128,
+             complete,
+             reset,
+         },
+     ]
+
+createLowLevelHasher : { seed ?U64 } -> LowLevelHasher
+createLowLevelHasher = \{ seed ? 0x526F_6352_616E_643F } ->
+    @LowLevelHasher { originalSeed: seed, state: seed }
+
+combineState : LowLevelHasher, { a : U64, b : U64, seed : U64, length : U64 } -> LowLevelHasher
+combineState = \@LowLevelHasher { originalSeed, state }, { a, b, seed, length } ->
+    tmp = wymix (Num.bitwiseXor wyp1 a) (Num.bitwiseXor seed b)
+    hash = wymix (Num.bitwiseXor wyp1 length) tmp
+
+    @LowLevelHasher { originalSeed, state: wymix state hash }
+
+complete = \@LowLevelHasher { state } -> state
+
+reset = \@LowLevelHasher { originalSeed } ->
+    @LowLevelHasher { originalSeed, state: originalSeed }
+
+addI8 = \hasher, i8 ->
+    addU8 hasher (Num.toU8 i8)
+addI16 = \hasher, i16 ->
+    addU16 hasher (Num.toU16 i16)
+addI32 = \hasher, i32 ->
+    addU32 hasher (Num.toU32 i32)
+addI64 = \hasher, i64 ->
+    addU64 hasher (Num.toU64 i64)
+addI128 = \hasher, i128 ->
+    addU128 hasher (Num.toU128 i128)
```

Comment:
Do you think there will be hashers that do something other than convert the signed repr to the unsigned repr here? If not, should we just get rid of the `addI*` variants and keep only the unsigned ones?

Reply:
I would expect all hashers to do this. Hashing algorithms really just care about the underlying bytes.